Sources/Recycling/GridViews/Ads/GridView/AdsGridView.swift

```diff
@@ -183,8 +184,22 @@ extension AdsGridView: UICollectionViewDataSource {
 // MARK: - AdsGridViewCellDataSource
 
 extension AdsGridView: AdsGridViewCellDataSource {
+    public func adsGridViewCell(_ adsGridViewCell: AdsGridViewCell, cachedImageForModel model: AdsGridViewModel) -> UIImage? {
+        guard let imagePath = model.imagePath else {
+            return nil
+        }
+
+        return imageCache.image(forKey: imagePath)
+    }
+
     public func adsGridViewCell(_ adsGridViewCell: AdsGridViewCell, loadImageForModel model: AdsGridViewModel, imageWidth: CGFloat, completion: @escaping ((AdsGridViewModel, UIImage?) -> Void)) {
-        dataSource?.adsGridView(self, loadImageForModel: model, imageWidth: imageWidth, completion: completion)
+        dataSource?.adsGridView(self, loadImageForModel: model, imageWidth: imageWidth, completion: { [weak self] model, image in
+            if let image = image, let imagePath = model.imagePath {
+                self?.imageCache.add(image, forKey: imagePath)
```

Comment:
Don't we have some sort of image caching in the app?

Reply:
Yes, I hope this would already be cached when used in the app

Reply:
The network layer already caches the images if that's what you're thinking about. The problem I'm really solving here is that the cell needs to know whether the image was already loaded _(and as such we don't need to animate)_, or whether the image was fetched from the internet _(and was therefore probably not previously visible)_.

Reply:
Aha, now I understand. So basically you want a way to check if the image should be set with animation... 🤔 

And you use a weak reference to the image so if it is expelled from other cache it would be nilled here too, so maybe all good then 🤔  

What do you think @vadymmarkov?

Reply:
Yeah, we should be good I think. Then there's no need for using `NSCache` instead of dictionary @pimms